test/unit/org/apache/cassandra/service/reads/range/RangeIteratorTest.java

```diff
@@ -236,4 +181,71 @@ public void testGRRExactBounds() throws Throwable
         testGRRKeys(exBounds(rp(""), rp("")), range(rp(""), endOf("1")), range(endOf("1"), endOf("6")), exBounds(endOf("6"), rp("")));
         testGRRKeys(incExBounds(rp(""), rp("")), bounds(rp(""), endOf("1")), range(endOf("1"), endOf("6")), exBounds(endOf("6"), rp("")));
     }
+
+    /**
+     * Test {@link RangeIterator#getRestrictedRanges(AbstractBounds)} for tokens.
+     */
+    @SafeVarargs
+    private final void testGRR(AbstractBounds<Token> queryRange, AbstractBounds<Token>... expected)
+    {
+        // Testing for tokens
+        List<AbstractBounds<Token>> restricted = RangeIterator.getRestrictedRanges(queryRange);
```

Comment:
It's kind of funny that `RangeIteratorTest` never actually creates a `RangeIterator` instance. It might make sense to change that and test the public contract of the class rather than reach into `getRestrictedRanges()`. (There is a little bit of conditional logic and iterator logic we miss too, of course.)

Reply:
What I don't understand is why `getRestrictedRanges` accepts any kind of `RingPosition` bounds when it's only invoked with bounds of `PartitionPosition`s obtained exclusively from `DataRange#keyRange`.

The test takes care of testing bounds of `Token`s but we never use those, so I wonder whether we could just get rid of the generics in `getRestrictedRanges` and the associated tests for tokens. WDYT?

Also, we could add a getter for the ranges and make `getRestrictedRanges` private, so we tests that attribute instead of the static method. Unfortunately testing `computeNext` seems more difficult since it tries to contact nodes that aren't really there.

Reply:
[Here](https://github.com/apache/cassandra/pull/801/commits/4ca3e0321a99af6d7457134fe5832bcb2c0621da) I have tried to remove the support for `AbstractBounds<Token>` in `getRestrictedRanges`, that simplifies both that function and the tests. I think that was a remainder from quite older versions.

[Here](https://github.com/apache/cassandra/pull/801/commits/237df2321f26220dddb15a52d3a88cc6b5917b8e) I have modified the test to actually instantiate `ReplicaPlanIterator` and verify the created ranges, including the special case of local replication strategy. The only part that is not covered is the actual building of the `ReplicaPlan` for each range ([this line](https://github.com/apache/cassandra/blob/237df2321f26220dddb15a52d3a88cc6b5917b8e/src/java/org/apache/cassandra/service/reads/range/ReplicaPlanIterator.java#L73)), since I think that it would require to have the replica nodes running.

Reply:
@adelapena I've read through the changes, and they all look good so far.